core/templates/pages/exploration-editor-page/services/exploration-states.service.ts

```diff
@@ -18,571 +18,759 @@
-
-    var stateAddedCallbacks = [];
-    var stateDeletedCallbacks = [];
-    var stateRenamedCallbacks = [];
-    var stateInteractionSavedCallbacks = [];
-    /** @private */
-    var refreshGraphEventEmitter = new EventEmitter();
-
-    // Properties that have a different backend representation from the
-    // frontend and must be converted.
-    var BACKEND_CONVERSIONS = {
-      answer_groups: function(answerGroups) {
-        return answerGroups.map(function(answerGroup) {
-          return answerGroup.toBackendDict();
-        });
-      },
-      content: function(content) {
-        return content.toBackendDict();
-      },
-      recorded_voiceovers: function(recordedVoiceovers) {
-        return recordedVoiceovers.toBackendDict();
-      },
-      default_outcome: function(defaultOutcome) {
-        if (defaultOutcome) {
-          return defaultOutcome.toBackendDict();
-        } else {
-          return null;
-        }
-      },
-      hints: function(hints) {
-        return hints.map(function(hint) {
-          return hint.toBackendDict();
-        });
-      },
-      param_changes: function(paramChanges) {
-        return paramChanges.map(function(paramChange) {
-          return paramChange.toBackendDict();
-        });
-      },
-      param_specs: function(paramSpecs) {
-        return paramSpecs.toBackendDict();
-      },
-      solution: function(solution) {
-        if (solution) {
-          return solution.toBackendDict();
-        } else {
-          return null;
-        }
-      },
-      written_translations: function(writtenTranslations) {
-        return writtenTranslations.toBackendDict();
-      },
-      widget_customization_args: function(customizationArgs) {
-        return Interaction.convertCustomizationArgsToBackendDict(
-          customizationArgs);
+@Injectable({
+  providedIn: 'root'
+})
+export class ExplorationStatesService {
+  constructor(
+    private contextService: ContextService,
+    private changeListService: ChangeListService,
+    private statesObjectFactory: StatesObjectFactory,
+    private solutionValidityService: SolutionValidityService,
+    private answerClassificationService: AnswerClassificationService,
+    private angularNameService: AngularNameService,
+    private alertsService: AlertsService,
+    private validatorsService: ValidatorsService,
+    private explorationInitStateNameService: ExplorationInitStateNameService,
+    private ngbModal: NgbModal,
+    private stateEditorService: StateEditorService,
+    private stateEditorRefreshService: StateEditorRefreshService,
+    private normalizeWhitespacePipe: NormalizeWhitespacePipe,
+    private location: Location,
+    private interactionRulesRegistryService: InteractionRulesRegistryService
+  ) {}
+
+  stateAddedCallbacks = [];
+  stateDeletedCallbacks = [];
+  stateRenamedCallbacks = [];
+  stateInteractionSavedCallbacks = [];
+  private _states: States = null;
+  private _refreshGraphEventEmitter = new EventEmitter();
```

Comment:
please shift these properties above the constructor and add appropriate types.

Reply:
Done.